File: Top-Interview-Questions/Strings/Easy-Collection/14-longest-common-prefix.py
class Solution:
    # Sol 1. Brute Force
    def longestCommonPrefix(self, strs: List[str]) -> str: # inputs are all lower case a-z, 26 letters
        ret_str = ""
        if len(strs) < 1:
            return ret_str

        prefix_str = min(strs, key=len)
        for i in range(len(prefix_str)):
            for s in strs:
                if s[i]!=prefix_str[i]:
                    return prefix_str[:i] 
                      
        return prefix_str
    
    # TC: O(n), worst case: prefix of a entire string
    # SC: O(n), splitting string uses extra memory
    # Runtime: 32 ms, faster than 74.43%
    # Memory: 13.9 MB, less than 6.67% 

File: Top-Interview-Questions/Strings/Easy-Collection/14-longest-common-prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 * Runtime: 76 ms, faster than 23.94% of JavaScript online submissions for Longest Common Prefix.
 * Memory Usage: 39.1 MB, less than 6.25% of JavaScript online submissions for Longest Common Prefix.
 * TC: O(LN), where L is the average length of strings in strs, and M is number of elements in strs
 * SC: O(1), assuming we're not counting output
 */
var longestCommonPrefix = function(strs) {
    if (!strs || strs.length === 0) {
        return "";
    } else if (strs.length === 1) {
        return strs[0];
    }
    
    const result = [];
    const pivot = strs[0];
    for (let index = 0; index < pivot.length; index++) {
        const candidate = pivot.charAt(index);
        
        for (const str of strs) {
            if (candidate !== str.charAt(index)) {
                return result.join("");
            }
        }
        
        result.push(candidate);
    }
    
    return result.join("");
};